java/ql/test/TestUtilities/JavaInlineExpectations.ql

```diff
@@ -0,0 +1,88 @@
+import java
+
+external predicate queryResults(string relation, int row, int column, string data);
+
+string actualLines() {
+  exists(int i |
+    queryResults("#select", i, _, _) and
+    result =
+      " | " +
+        concat(int j, string cell | queryResults("#select", i, j, cell) | cell, " | " order by j) +
+        " | "
+  )
+}
+
+predicate parsedExpectedResults(string filename, int line, string content) {
+  exists(Javadoc doc |
+    isEolComment(doc) and
+    filename = doc.getLocation().getFile().getBaseName() and
+    line = doc.getLocation().getStartLine() and
+    line = doc.getLocation().getEndLine() and
+    content = doc.getChild(0).getText().trim()
+  )
+}
+
+predicate expectError(string filename, int line) {
+  exists(string content |
+    parsedExpectedResults(filename, line, content) and content.indexOf("NOT OK") = 0
+  )
+}
+
+predicate expectPass(string filename, int line) {
+  exists(string content |
+    parsedExpectedResults(filename, line, content) and content.indexOf("OK") = 0
+  )
+}
+
+predicate parsedActualResults(string filename, int line, int colStart, int colEnd, string content) {
+  exists(string s, string posString, string lineString |
+    s = actualLines() and
+    posString = s.substring(s.indexOf("|", 0, 0) + 1, s.indexOf("|", 1, 0)).trim() and
+    filename = posString.substring(0, posString.indexOf(":", 0, 0)) and
+    lineString = posString.substring(posString.indexOf(":", 0, 0) + 1, posString.indexOf(":", 1, 0)) and
+    lineString = posString.substring(posString.indexOf(":", 2, 0) + 1, posString.indexOf(":", 3, 0)) and
+    colStart =
+      posString.substring(posString.indexOf(":", 1, 0) + 1, posString.indexOf(":", 2, 0)).toInt() and
+    colEnd = posString.substring(posString.indexOf(":", 3, 0) + 1, posString.length()).toInt() and
+    line = lineString.toInt() and
+    content = s.substring(s.indexOf("|", 2, 0) + 1, s.indexOf("|", 3, 0)).trim()
```

Comment:
Is this meant as an example of best practice? If not, I'd like to see such an example.

Reply:
What I mean is that the QL code works by concatenating all the columns and then splitting by `|` to take them apart again. I hope that's unnecessary.

Reply:
Apologies, I see what you mean now.

This concatenation is completely unnecessary and merely an artifact of me having gone through multiple different interfaces while working on this issue and not properly cleaning up the QL code after the most recent change.

I have pushed a third commit to remove `actualLines` and with it the concatenation.